# Data Cleaning
> Objectives:
    - Given a messy dataset - extract meaningful information from it. 
    - Learn from the data.

In [75]:
import pandas as pd
import numpy as np

In [ ]:
# read vehicles_messy.csv
# quick analysis
# display more columns on the dataframe visualization

## Data types

### Checking data types of our dataset

In [76]:
pd.set_option('display.max_columns',None)

In [4]:
pd.set_option('display.max_rows',60)

In [6]:
# Your code here!
df = pd.read_csv("vehicles_messy.csv", dtype={'year':'object'})

/var/folders/4s/j3x9y61j1gj2zdqsyf0tqv0w0000gn/T/ipykernel_3412/3550249580.py:2: DtypeWarning: Columns (70,71,72,73,74,76,79) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("vehicles_messy.csv", dtype={'year':'object'})


In [77]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37843 entries, 0 to 37842
Data columns (total 83 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   barrels08        37843 non-null  float64
 1   barrelsA08       37843 non-null  float64
 2   charge120        37843 non-null  float64
 3   charge240        37843 non-null  float64
 4   city08           37843 non-null  int64  
 5   city08U          37843 non-null  float64
 6   cityA08          37843 non-null  int64  
 7   cityA08U         37843 non-null  float64
 8   cityCD           37843 non-null  float64
 9   cityE            37843 non-null  float64
 10  cityUF           37843 non-null  float64
 11  co2              37843 non-null  int64  
 12  co2A             37843 non-null  int64  
 13  co2TailpipeAGpm  37843 non-null  float64
 14  co2TailpipeGpm   37843 non-null  float64
 15  comb08           37843 non-null  int64  
 16  comb08U          37843 non-null  float64
 17  combA08     

In [14]:
type(df.dtypes)

pandas.core.series.Series

In [20]:
df.select_dtypes(include='object').head()

drive    eng_dscr fuelType         fuelType1  \
0            Rear-Wheel Drive       (FFS)  Regular  Regular Gasoline   
1            Rear-Wheel Drive   (GUZZLER)  Regular  Regular Gasoline   
2           Front-Wheel Drive       (FFS)  Regular  Regular Gasoline   
3            Rear-Wheel Drive         NaN  Regular  Regular Gasoline   
4  4-Wheel or All-Wheel Drive  (FFS,TRBO)  Premium  Premium Gasoline   

         make                model mpgData            trany           VClass  \
0  Alfa Romeo   Spider Veloce 2000       Y     Manual 5-spd      Two Seaters   
1     Ferrari           Testarossa       N     Manual 5-spd      Two Seaters   
2       Dodge              Charger       Y     Manual 5-spd  Subcompact Cars   
3       Dodge  B150/B250 Wagon 2WD       N  Automatic 3-spd             Vans   
4      Subaru     Legacy AWD Turbo       N     Manual 5-spd     Compact Cars   

   year guzzler trans_dscr tCharger sCharger atvType fuelType2 rangeA evMotor  \
0  1985     NaN        NaN      NaN      NaN     NaN       NaN    NaN     NaN   
1  1985       T        NaN      NaN      NaN     NaN       NaN    NaN     NaN   
2  1985     NaN        SIL      NaN      NaN     NaN       NaN    NaN     NaN   
3  1985     NaN        NaN      NaN      NaN     NaN       NaN    NaN     NaN   
4  1993     NaN        NaN        T      NaN     NaN       NaN    NaN     NaN   

  mfrCode c240Dscr c240bDscr                     createdOn  \
0     NaN      NaN       NaN  Tue Jan 01 00:00:00 EST 2013   
1     NaN      NaN       NaN  Tue Jan 01 00:00:00 EST 2013   
2     NaN      NaN       NaN  Tue Jan 01 00:00:00 EST 2013   
3     NaN      NaN       NaN  Tue Jan 01 00:00:00 EST 2013   
4     NaN      NaN       NaN  Tue Jan 01 00:00:00 EST 2013   

                     modifiedOn startStop  
0  Tue Jan 01 00:00:00 EST 2013       NaN  
1  Tue Jan 01 00:00:00 EST 2013       NaN  
2  Tue Jan 01 00:00:00 EST 2013       NaN  
3  Tue Jan 01 00:00:00 EST 2013       NaN  
4  Tue Jan 01 00:00:00 EST 2013       NaN

In [19]:
df.select_dtypes(include='object').info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37843 entries, 0 to 37842
Data columns (total 24 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   drive       36654 non-null  object
 1   eng_dscr    22440 non-null  object
 2   fuelType    37843 non-null  object
 3   fuelType1   37843 non-null  object
 4   make        37843 non-null  object
 5   model       37843 non-null  object
 6   mpgData     37843 non-null  object
 7   trany       37832 non-null  object
 8   VClass      37843 non-null  object
 9   year        37843 non-null  object
 10  guzzler     2281 non-null   object
 11  trans_dscr  15047 non-null  object
 12  tCharger    5186 non-null   object
 13  sCharger    666 non-null    object
 14  atvType     3072 non-null   object
 15  fuelType2   1408 non-null   object
 16  rangeA      1403 non-null   object
 17  evMotor     562 non-null    object
 18  mfrCode     7025 non-null   object
 19  c240Dscr    37 non-null     object
 20  c240bD

In [27]:
df.dtypes[df.dtypes == 'object']

drive         object
eng_dscr      object
fuelType      object
fuelType1     object
make          object
model         object
mpgData       object
trany         object
VClass        object
guzzler       object
trans_dscr    object
tCharger      object
sCharger      object
atvType       object
fuelType2     object
rangeA        object
evMotor       object
mfrCode       object
c240Dscr      object
c240bDscr     object
createdOn     object
modifiedOn    object
startStop     object
dtype: object

### Convert values using `.astype()`

Converts the type of the column

In [ ]:
df['year'] = df['year'].astype('int')

### Convert values using map

In [28]:
df['year'] = df['year'].astype('str')

In [29]:
df['year'].map(int) ## atenção: MAP transforma qualquer função em "metodo" de pandas DataFrame

0        1985
1        1985
2        1985
3        1985
4        1993
         ... 
37838    1993
37839    1993
37840    1993
37841    1993
37842    1993
Name: year, Length: 37843, dtype: int64

In [38]:
np.nan

nan

In [45]:
def int_forcado(ano):
    try:
        return int(ano)
    except:
        return np.nan

In [ ]:
['1985', '1986', '1987', 'Renata', '1988']

In [40]:
df_year = pd.DataFrame(['1985', '1986', '1987', 'Renata', '1988'], columns=['year'])

In [41]:
df_year

year
0    1985
1    1986
2    1987
3  Renata
4    1988

In [48]:
# df_year.astype('int')
df_year['year'].map(int_forcado)

0    1985.0
1    1986.0
2    1987.0
3       NaN
4    1988.0
Name: year, dtype: float64

In [33]:
def classif_ano(ano):
    if int(ano) <= 1990:
        return 'A'
    else:
        return 'B'

In [34]:
classif_ano(1885)

'A'

In [35]:
df['year_classif'] = df['year'].map(classif_ano)

0        A
1        A
2        A
3        A
4        B
        ..
37838    B
37839    B
37840    B
37841    B
37842    B
Name: year, Length: 37843, dtype: object

## Null (or missing) values

>    - Called NaN: <u>Not A Number</u>
>    - Count 
>    - <b>Mask</b> concept

### `np.nan`

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37843 entries, 0 to 37842
Data columns (total 83 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   barrels08        37843 non-null  float64
 1   barrelsA08       37843 non-null  float64
 2   charge120        37843 non-null  float64
 3   charge240        37843 non-null  float64
 4   city08           37843 non-null  int64  
 5   city08U          37843 non-null  float64
 6   cityA08          37843 non-null  int64  
 7   cityA08U         37843 non-null  float64
 8   cityCD           37843 non-null  float64
 9   cityE            37843 non-null  float64
 10  cityUF           37843 non-null  float64
 11  co2              37843 non-null  int64  
 12  co2A             37843 non-null  int64  
 13  co2TailpipeAGpm  37843 non-null  float64
 14  co2TailpipeGpm   37843 non-null  float64
 15  comb08           37843 non-null  int64  
 16  comb08U          37843 non-null  float64
 17  combA08     

In [68]:
df.sCharger.isna().value_counts()

True     37177
False      666
Name: sCharger, dtype: int64

In [70]:
df.isna().groupby('sCharger')['sCharger'].count()

sCharger
False      666
True     37177
Name: sCharger, dtype: int64

In [60]:
df.isna()

barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0          False       False      False      False   False    False    False   
1          False       False      False      False   False    False    False   
2          False       False      False      False   False    False    False   
3          False       False      False      False   False    False    False   
4          False       False      False      False   False    False    False   
...          ...         ...        ...        ...     ...      ...      ...   
37838      False       False      False      False   False    False    False   
37839      False       False      False      False   False    False    False   
37840      False       False      False      False   False    False    False   
37841      False       False      False      False   False    False    False   
37842      False       False      False      False   False    False    False   

       cityA08U  cityCD  cityE  cityUF    co2   co2A  co2TailpipeAGpm  \
0         False   False  False   False  False  False            False   
1         False   False  False   False  False  False            False   
2         False   False  False   False  False  False            False   
3         False   False  False   False  False  False            False   
4         False   False  False   False  False  False            False   
...         ...     ...    ...     ...    ...    ...              ...   
37838     False   False  False   False  False  False            False   
37839     False   False  False   False  False  False            False   
37840     False   False  False   False  False  False            False   
37841     False   False  False   False  False  False            False   
37842     False   False  False   False  False  False            False   

       co2TailpipeGpm  comb08  comb08U  combA08  combA08U  combE  combinedCD  \
0               False   False    False    False     False  False       False   
1               False   False    False    False     False  False       False   
2               False   False    False    False     False  False       False   
3               False   False    False    False     False  False       False   
4               False   False    False    False     False  False       False   
...               ...     ...      ...      ...       ...    ...         ...   
37838           False   False    False    False     False  False       False   
37839           False   False    False    False     False  False       False   
37840           False   False    False    False     False  False       False   
37841           False   False    False    False     False  False       False   
37842           False   False    False    False     False  False       False   

       combinedUF  cylinders  displ  drive  engId  eng_dscr  feScore  \
0           False      False  False  False  False     False    False   
1           False      False  False  False  False     False    False   
2           False      False  False  False  False     False    False   
3           False      False  False  False  False      True    False   
4           False      False  False  False  False     False    False   
...           ...        ...    ...    ...    ...       ...      ...   
37838       False      False  False  False  False     False    False   
37839       False      False  False  False  False     False    False   
37840       False      False  False  False  False     False    False   
37841       False      False  False  False  False     False    False   
37842       False      False  False  False  False     False    False   

       fuelCost08  fuelCostA08  fuelType  fuelType1  ghgScore  ghgScoreA  \
0           False        False     False      False     False      False   
1           False        False     False      False     False      False   
2           False        False     False      False     False      False   
3           False        False     Fals

In [55]:
df.sCharger.notnull()

0        False
1        False
2        False
3        False
4        False
         ...  
37838    False
37839    False
37840    False
37841    False
37842    False
Name: sCharger, Length: 37843, dtype: bool

In [49]:
type(np.nan)

float

### Let's select the displ column and see how many missing values this column has.

### How to count it how many missing there are?

In [ ]:
#isna() or isnull() as above

In [91]:
(df.isna().sum()*100/df.shape[0]).sort_values(ascending=False).astype('int')

c240bDscr          99
c240Dscr           99
evMotor            98
sCharger           98
rangeA             96
fuelType2          96
guzzler            93
atvType            91
tCharger           86
startStop          83
mfrCode            81
trans_dscr         60
eng_dscr           40
drive               3
cylinders           0
displ               0
trany               0
rangeCity           0
rangeCityA          0
rangeHwy            0
barrels08           0
range               0
pv4                 0
pv2                 0
phevBlended         0
mpgData             0
model               0
make                0
rangeHwyA           0
VClass              0
UCity               0
UCityA              0
UHighway            0
UHighwayA           0
lv2                 0
year                0
youSaveSpend        0
charge240b          0
createdOn           0
modifiedOn          0
phevCity            0
phevHwy             0
lv4                 0
hlv                 0
id                  0
cityUF    

In [21]:
serie_nulos = (df.isna().sum()*100/df.shape[0]).sort_values(ascending=False).astype('int')
((serie_nulos > 0).value_counts(normalize=True))

False    0.831325
True     0.168675
dtype: float64

In [22]:
83*0.168675

14.000024999999999

### What if we wanted to count Null values for each row?
    - axis = 1 
    


In [94]:
df.shape

(37843, 83)

In [96]:
df.head()

barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0  15.695714         0.0        0.0        0.0      19      0.0        0   
1  29.964545         0.0        0.0        0.0       9      0.0        0   
2  12.207778         0.0        0.0        0.0      23      0.0        0   
3  29.964545         0.0        0.0        0.0      10      0.0        0   
4  17.347895         0.0        0.0        0.0      17      0.0        0   

   cityA08U  cityCD  cityE  cityUF  co2  co2A  co2TailpipeAGpm  \
0       0.0     0.0    0.0     0.0   -1    -1              0.0   
1       0.0     0.0    0.0     0.0   -1    -1              0.0   
2       0.0     0.0    0.0     0.0   -1    -1              0.0   
3       0.0     0.0    0.0     0.0   -1    -1              0.0   
4       0.0     0.0    0.0     0.0   -1    -1              0.0   

   co2TailpipeGpm  comb08  comb08U  combA08  combA08U  combE  combinedCD  \
0      423.190476      21      0.0        0       0.0    0.0         0.0   
1      807.909091      11      0.0        0       0.0    0.0         0.0   
2      329.148148      27      0.0        0       0.0    0.0         0.0   
3      807.909091      11      0.0        0       0.0    0.0         0.0   
4      467.736842      19      0.0        0       0.0    0.0         0.0   

   combinedUF  cylinders  displ                       drive  engId  \
0         0.0        4.0    2.0            Rear-Wheel Drive   9011   
1         0.0       12.0    4.9            Rear-Wheel Drive  22020   
2         0.0        4.0    2.2           Front-Wheel Drive   2100   
3         0.0        8.0    5.2            Rear-Wheel Drive   2850   
4         0.0        4.0    2.2  4-Wheel or All-Wheel Drive  66031   

     eng_dscr  feScore  fuelCost08  fuelCostA08 fuelType         fuelType1  \
0       (FFS)       -1        1600            0  Regular  Regular Gasoline   
1   (GUZZLER)       -1        3050            0  Regular  Regular Gasoline   
2       (FFS)       -1        1250            0  Regular  Regular Gasoline   
3         NaN       -1        3050            0  Regular  Regular Gasoline   
4  (FFS,TRBO)       -1        2150            0  Premium  Premium Gasoline   

   ghgScore  ghgScoreA  highway08  highway08U  highwayA08  highwayA08U  \
0        -1         -1         25         0.0           0          0.0   
1        -1         -1         14         0.0           0          0.0   
2        -1         -1         33         0.0           0          0.0   
3        -1         -1         12         0.0           0          0.0   
4        -1         -1         23         0.0           0          0.0   

   highwayCD  highwayE  highwayUF  hlv  hpv     id  lv2  lv4        make  \
0        0.0       0.0        0.0    0    0      1    0    0  Alfa Romeo   
1        0.0       0.0        0.0    0    0     10    0    0     Ferrari   
2        0.0       0.0        0.0   19   77    100    0    0       Dodge   
3        0.0       0.0        0.0    0    0   1000    0    0       Dodge   
4        0.0       0.0        0.0    0    0  10000    0   14      Subaru   

                 model mpgData  phevBlended  pv2  pv4  range  rangeCity  \
0   Spider Veloce 2000       Y        False    0    0      0        0.0   
1           Testarossa       N        False    0    0      0        0.0   
2              Charger       Y        False    0    0      0        0.0   
3  B150/B250 Wagon 2WD       N        False    0    0      0        0.0   
4     Legacy AWD Turbo       N        False    0   90      0        0.0   

   rangeCityA  rangeHwy  rangeHwyA            trany    UCity  UCityA  \
0         0.0       0.0        0.0     Manual 5-spd  23.3333     0.0   
1         0.0       0.0        0.0     Manual 5-spd  11.0000     0.0   
2         0.0       0.0        0.0     Manual 5-spd  29.0000     0.0   
3         0.0       0.0        0.0  Automatic 3-spd  12.2222     0.0   
4         0.0       0.0        0.0     Manual 5-spd  21.0000     0.0   

   UHighway  UHi

In [14]:
pd.set_option('display.max_rows', None)

In [16]:
df.isna().sum(axis=1).sort_values(ascending=False).head(30)

18374    14
18812    14
8144     14
18822    14
18541    14
18540    14
7138     14
18235    14
18696    14
18814    14
18236    14
18912    14
21413    14
18885    14
18709    14
19450    14
18888    14
18512    14
18914    14
18516    14
18623    14
18513    14
18543    14
18621    14
18824    14
18828    14
18225    14
18349    14
18227    14
8147     14
dtype: int64

### What if I wanted to select only the rows in which there are at least one missing value (in any column)?

In [26]:
df.isna().sum(axis=1).value_counts(normalize=True).sort_index()

4     0.000053
5     0.000925
6     0.002352
7     0.013054
8     0.016859
9     0.073541
10    0.081204
11    0.282800
12    0.398066
13    0.130196
14    0.000951
dtype: float64

## Dropping columns
> `axis=1` or

> `columns = ['name_of_column_to_drop1', 'name_of_column_to_drop2' ,...]`

### Drop columns based on condition

In [34]:
col_null_perc = df.isna().sum()*100/df.shape[0]
cols2drop = col_null_perc[col_null_perc > 40].index
cols2drop

Index(['eng_dscr', 'guzzler', 'trans_dscr', 'tCharger', 'sCharger', 'atvType',
       'fuelType2', 'rangeA', 'evMotor', 'mfrCode', 'c240Dscr', 'c240bDscr',
       'startStop'],
      dtype='object')

In [36]:
df_notnulls = df.drop(columns=cols2drop)
df_notnulls.shape

(37843, 70)

In [45]:
df_notnulls.isna().sum(axis=1).value_counts().sort_values(ascending=False).head(20)

0    36537
1     1186
2      106
3       11
4        3
dtype: int64

In [ ]:
df_nulllines = df.drop(index=lst_index2drop)

In [ ]:
df_notnulls.description

In [37]:
dir(df_notnulls)

['T',
 'UCity',
 'UCityA',
 'UHighway',
 'UHighwayA',
 'VClass',
 '_AXIS_LEN',
 '_AXIS_ORDERS',
 '_AXIS_TO_AXIS_NUMBER',
 '_HANDLED_TYPES',
 '__abs__',
 '__add__',
 '__and__',
 '__annotations__',
 '__array__',
 '__array_priority__',
 '__array_ufunc__',
 '__array_wrap__',
 '__bool__',
 '__class__',
 '__contains__',
 '__copy__',
 '__deepcopy__',
 '__delattr__',
 '__delitem__',
 '__dict__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__finalize__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__iand__',
 '__ifloordiv__',
 '__imod__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__invert__',
 '__ior__',
 '__ipow__',
 '__isub__',
 '__iter__',
 '__itruediv__',
 '__ixor__',
 '__le__',
 '__len__',
 '__lt__',
 '__matmul__',
 '__mod__',
 '__module__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__nonzero__',
 '__or__',
 '__pos__',
 '__pow__',
 '__radd__',
 '__rand__',
 '__

### Using `dropna`

In [39]:
df.dropna().shape

(0, 83)

In [41]:
df_notnulls.shape

(37843, 70)

In [40]:
df_notnulls.dropna().shape

(36537, 70)

In [44]:
(1 - df_notnulls.dropna().shape[0]/df_notnulls.shape[0])*100

3.451100599846735

In [ ]:
df_notnulls.dropna(how='all')

In [58]:
df.shape[0]*0.60

22705.8

In [57]:
df_notnulls2 = df.dropna(axis=1, thresh=df.shape[0]*0.60)
df_notnulls2.shape

(37843, 70)

In [ ]:
#equiv
col_null_perc = df.isna().sum()*100/df.shape[0]
cols2drop = col_null_perc[col_null_perc > 40].index
cols2drop
df_notnulls = df.drop(columns=cols2drop)

### Fill with some value

In [63]:
df_notnulls.isna().sum().sort_values(ascending=False).head()

drive        1189
cylinders     123
displ         120
trany          11
barrels08       0
dtype: int64

In [ ]:
df_notnulls.v

In [66]:
df_notnulls.head()

barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0  15.695714         0.0        0.0        0.0      19      0.0        0   
1  29.964545         0.0        0.0        0.0       9      0.0        0   
2  12.207778         0.0        0.0        0.0      23      0.0        0   
3  29.964545         0.0        0.0        0.0      10      0.0        0   
4  17.347895         0.0        0.0        0.0      17      0.0        0   

   cityA08U  cityCD  cityE  cityUF  co2  co2A  co2TailpipeAGpm  \
0       0.0     0.0    0.0     0.0   -1    -1              0.0   
1       0.0     0.0    0.0     0.0   -1    -1              0.0   
2       0.0     0.0    0.0     0.0   -1    -1              0.0   
3       0.0     0.0    0.0     0.0   -1    -1              0.0   
4       0.0     0.0    0.0     0.0   -1    -1              0.0   

   co2TailpipeGpm  comb08  comb08U  combA08  combA08U  combE  combinedCD  \
0      423.190476      21      0.0        0       0.0    0.0         0.0   
1      807.909091      11      0.0        0       0.0    0.0         0.0   
2      329.148148      27      0.0        0       0.0    0.0         0.0   
3      807.909091      11      0.0        0       0.0    0.0         0.0   
4      467.736842      19      0.0        0       0.0    0.0         0.0   

   combinedUF  cylinders  displ                       drive  engId  feScore  \
0         0.0        4.0    2.0            Rear-Wheel Drive   9011       -1   
1         0.0       12.0    4.9            Rear-Wheel Drive  22020       -1   
2         0.0        4.0    2.2           Front-Wheel Drive   2100       -1   
3         0.0        8.0    5.2            Rear-Wheel Drive   2850       -1   
4         0.0        4.0    2.2  4-Wheel or All-Wheel Drive  66031       -1   

   fuelCost08  fuelCostA08 fuelType         fuelType1  ghgScore  ghgScoreA  \
0        1600            0  Regular  Regular Gasoline        -1         -1   
1        3050            0  Regular  Regular Gasoline        -1         -1   
2        1250            0  Regular  Regular Gasoline        -1         -1   
3        3050            0  Regular  Regular Gasoline        -1         -1   
4        2150            0  Premium  Premium Gasoline        -1         -1   

   highway08  highway08U  highwayA08  highwayA08U  highwayCD  highwayE  \
0         25         0.0           0          0.0        0.0       0.0   
1         14         0.0           0          0.0        0.0       0.0   
2         33         0.0           0          0.0        0.0       0.0   
3         12         0.0           0          0.0        0.0       0.0   
4         23         0.0           0          0.0        0.0       0.0   

   highwayUF  hlv  hpv     id  lv2  lv4        make                model  \
0        0.0    0    0      1    0    0  Alfa Romeo   Spider Veloce 2000   
1        0.0    0    0     10    0    0     Ferrari           Testarossa   
2        0.0   19   77    100    0    0       Dodge              Charger   
3        0.0    0    0   1000    0    0       Dodge  B150/B250 Wagon 2WD   
4        0.0    0    0  10000    0   14      Subaru     Legacy AWD Turbo   

  mpgData  phevBlended  pv2  pv4  range  rangeCity  rangeCityA  rangeHwy  \
0       Y        False    0    0      0        0.0         0.0       0.0   
1       N        False    0    0      0        0.0         0.0       0.0   
2       Y        False    0    0      0        0.0         0.0       0.0   
3       N        False    0    0      0        0.0         0.0       0.0   
4       N        False    0   90      0        0.0         0.0       0.0   

   rangeHwyA            trany    UCity  UCityA  UHighway  UHighwayA  \
0        0.0     Manual 5-spd  23.3333     0.0   35.0000        0.0   
1        0.0     Manual 5-spd  11.0000     0.0   19.0000        0.0   
2        0.0     Manual 5-spd  29.0000     0.0   47.0000        0.0   
3        0.0  Automatic 3-spd  12.2222     0.0   16.6667        0.0   
4        0.0     Manual 5-spd  21.

In [69]:
df.co2.value_counts()

-1      31989
 0        120
 413       66
 382       41
 344       41
 358       41
 354       41
 374       41
 425       39
 317       38
 301       37
 386       36
 475       35
 521       34
 357       34
 405       33
 444       33
 323       33
 385       32
 295       32
 343       32
 340       32
 408       31
 523       31
 375       31
 346       31
 377       30
 324       30
 404       30
 349       30
 486       30
 288       30
 398       29
 432       29
 312       29
 313       29
 379       29
 517       28
 334       28
 465       28
 468       28
 481       28
 437       27
 372       27
 328       27
 397       27
 302       27
 458       27
 476       27
 373       27
 442       26
 407       26
 387       26
 470       26
 361       26
 337       25
 389       25
 376       25
 423       25
 424       25
 482       25
 370       25
 459       25
 392       25
 418       25
 426       24
 339       24
 297       24
 369       24
 314       24
 360       23
 410  

In [68]:
df_notnulls.co2.value_counts()

-1      31989
 0        120
 413       66
 382       41
 344       41
 358       41
 354       41
 374       41
 425       39
 317       38
 301       37
 386       36
 475       35
 521       34
 357       34
 405       33
 444       33
 323       33
 385       32
 295       32
 343       32
 340       32
 408       31
 523       31
 375       31
 346       31
 377       30
 324       30
 404       30
 349       30
 486       30
 288       30
 398       29
 432       29
 312       29
 313       29
 379       29
 517       28
 334       28
 465       28
 468       28
 481       28
 437       27
 372       27
 328       27
 397       27
 302       27
 458       27
 476       27
 373       27
 442       26
 407       26
 387       26
 470       26
 361       26
 337       25
 389       25
 376       25
 423       25
 424       25
 482       25
 370       25
 459       25
 392       25
 418       25
 426       24
 339       24
 297       24
 369       24
 314       24
 360       23
 410  

In [84]:
df_notnulls.isna().sum().sort_values(ascending=False).head()

drive        1189
cylinders     123
displ         120
trany          11
barrels08       0
dtype: int64

In [67]:
df_notnulls.drive.value_counts(normalize=True)

Front-Wheel Drive             0.362471
Rear-Wheel Drive              0.352786
4-Wheel or All-Wheel Drive    0.181372
All-Wheel Drive               0.056119
4-Wheel Drive                 0.029055
2-Wheel Drive                 0.013832
Part-time 4-Wheel Drive       0.004338
2-Wheel Drive, Front          0.000027
Name: drive, dtype: float64

In [83]:
df_notnulls.cylinders.value_counts(normalize=True).sort_index()

2.0     0.001432
3.0     0.005488
4.0     0.384040
5.0     0.020308
6.0     0.349523
8.0     0.219857
10.0    0.004056
12.0    0.015085
16.0    0.000212
Name: cylinders, dtype: float64

In [82]:
(df_notnulls.displ.value_counts(normalize=True)*100).sort_index()

0.0    0.005302
0.6    0.007953
0.9    0.010604
1.0    0.437399
1.1    0.031811
1.2    0.108687
1.3    0.498370
1.4    0.548737
1.5    1.887443
1.6    3.955147
1.7    0.201469
1.8    4.188426
1.9    1.108077
2.0    9.376243
2.1    0.280996
2.2    3.660897
2.3    2.982265
2.4    5.092384
2.5    6.272036
2.6    0.901307
2.7    1.293640
2.8    2.616441
2.9    0.524879
3.0    7.836068
3.1    0.675980
3.2    1.407629
3.3    1.129285
3.4    1.142539
3.5    3.859714
3.6    2.401718
3.7    1.489807
3.8    2.597885
3.9    0.827082
4.0    2.627045
4.1    0.251836
4.2    1.420884
4.3    3.631737
4.4    0.811176
4.5    0.238581
4.6    1.500411
4.7    0.988787
4.8    0.702489
4.9    0.933118
5.0    3.851762
5.2    1.447393
5.3    2.120722
5.4    0.816478
5.5    0.546086
5.6    0.469210
5.7    2.836466
5.8    0.424144
5.9    0.896005
6.0    1.110728
6.1    0.063622
6.2    1.699229
6.3    0.116640
6.4    0.082178
6.5    0.299552
6.6    0.053018
6.7    0.182912
6.8    0.368475
7.0    0.026509
7.4    0

1. Media de displ por Marca
2. 'Tesla', 'CODA Automotive', 'Azure Dynamics' tem TODOS os displ NaN
3. Essas marcasd representam 36 linhas
4. Porem temos 120 displ NaN (ou seja, temos outras marcas com displ NaN)
5. EURECA: Todos os 36 tem mesmo fuelType >> Electricity
6. Vimos que dos 120 eletricos(independente de marca), 118 estao Nulos e 2 zerados
7. Tinhamos 120 displ NaN, 118 são por serem eletricos, quais sao os outros 2?

In [85]:
df_notnulls.groupby(['make'])['displ'].mean()

make
AM General                            3.350000
ASC Incorporated                      3.800000
Acura                                 2.834768
Alfa Romeo                            2.528889
American Motors Corporation           3.696296
Aston Martin                          5.322535
Audi                                  2.850545
Aurora Cars Ltd                       4.900000
Autokraft Limited                     4.925000
Avanti Motor Corporation              5.000000
Azure Dynamics                             NaN
BMW                                   3.228580
BMW Alpina                            4.533333
BYD                                   0.000000
Bentley                               6.084483
Bertone                               1.500000
Bill Dovell Motor Car Company         2.300000
Bitter Gmbh and Co. Kg                3.360000
Bugatti                               8.000000
Buick                                 3.304127
CCC Engineering                       5.700000
CODA Aut

In [89]:
df_notnulls['make'].isin(['Tesla', 'CODA Automotive', 'Azure Dynamics']).value_counts()

False    37807
True        36
Name: make, dtype: int64

In [90]:
mask = df_notnulls['make'].isin(['Tesla', 'CODA Automotive', 'Azure Dynamics'])
df_notnulls[mask]

barrels08  barrelsA08  charge120  charge240  city08   city08U  cityA08  \
24051      0.324         0.0        0.0        8.0      62   61.9000        0   
24052      0.324         0.0        0.0        8.0      62   61.9000        0   
24473      0.276         0.0        0.0        6.0      77   77.2000        0   
24784      0.228         0.0        0.0       12.0      88   87.9000        0   
25650      0.276         0.0        0.0        6.0      77   77.2000        0   
25683      0.210         0.0        0.0       10.0      94   94.4000        0   
25684      0.228         0.0        0.0       12.0      88   87.9000        0   
25955      0.216         0.0        0.0        6.0      93   93.4000        0   
27226      0.228         0.0        0.0       12.0      88   87.9000        0   
27227      0.210         0.0        0.0       10.0      94   94.4000        0   
28560      0.228         0.0        0.0       12.0      88   87.9000        0   
28575      0.228         0.0        0.0       12.0      86   86.0000        0   
28591      0.216         0.0        0.0       12.0      89   89.3714        0   
28592      0.204         0.0        0.0       12.0      95   95.3056        0   
28600      0.210         0.0        0.0       10.0      94   94.4056        0   
28720      0.198         0.0        0.0       12.0     101  100.8098        0   
29443      0.204         0.0        0.0       12.0      95   95.3056        0   
29444      0.216         0.0        0.0       12.0      89   89.3714        0   
29652      0.222         0.0        0.0       12.0      90   90.2795        0   
29654      0.228         0.0        0.0       12.0      89   88.5914        0   
29933      0.228         0.0        0.0       10.0      88   87.8784        0   
29934      0.228         0.0        0.0       12.0      88   87.9000        0   
29935      0.228         0.0        0.0       12.0      88   87.8784        0   
29936      0.228         0.0        0.0       12.0      88   87.8784        0   
29938      0.198         0.0        0.0       12.0     101  100.8098        0   
29939      0.204         0.0        0.0       12.0      95   95.3056        0   
29941      0.198         0.0        0.0       12.0     101  100.8065        0   
29942      0.216         0.0        0.0       12.0      89   89.3714        0   
29943      0.210         0.0        0.0       12.0      91   91.3191        0   
30141      0.204         0.0        0.0       12.0      97   97.4898        0   
30142      0.198         0.0        0.0       12.0     102  101.7702        0   
30143      0.216         0.0        0.0       12.0      91   91.1628        0   
30972      0.204         0.0        0.0       10.0      98   97.5636        0   
30973      0.192         0.0        0.0       12.0     101  101.4750        0   
30974      0.210         0.0        0.0       12.0      92   92.4713        0   
30975      0.216         0.0        0.0       12.0      91   91.2362        0   

       cityA08U  cityCD  cityE  cityUF  co2  co2A  co2TailpipeAGpm  \
24051       0.0     0.0   55.0     0.0    0    -1              0.0   
24052       0.0     0.0   55.0     0.0    0    -1              0.0   
24473       0.0     0.0   44.0     0.0    0    -1              0.0   
24784       0.0     0.0   38.0     0.0    0    -1              0.0   
25650       0.0     0.0   44.0     0.0    0    -1              0.0   
25683       0.0     0.0   36.0     0.0    0    -1              0.0   
25684       0.0     0.0   38.0     0.0    0    -1              0.0   
25955       0.0     0.0   36.0     0.0    0    -1              0.0   
27226       0.0     0.0   38.0     0.0    0    -1              0.0   
27227       0.0     0.0   36.0     0.0    0    -1              0.0   
28560       0.0     0.0   38.0     0.0    0    -1              0.0   
28575       0.0     0.0   39.0     0.0    0    -1              0.0   
28591       0.0     0.0   38.0     0.0    0    -1              0.0   
28592       0.0     0.0   35.

In [94]:
df_notnulls.groupby('fuelType1')['fuelType1'].count()

fuelType1
Diesel                1101
Electricity            120
Midgrade Gasoline       74
Natural Gas             60
Premium Gasoline     10162
Regular Gasoline     26326
Name: fuelType1, dtype: int64

In [91]:
df_notnulls.groupby('fuelType')['fuelType'].count()

fuelType
CNG                               60
Diesel                          1101
Electricity                      120
Gasoline or E85                 1204
Gasoline or natural gas           20
Gasoline or propane                8
Midgrade                          74
Premium                        10004
Premium Gas or Electricity        17
Premium and Electricity           20
Premium or E85                   121
Regular                        25076
Regular Gas and Electricity       16
Regular Gas or Electricity         2
Name: fuelType, dtype: int64

In [98]:
df_elet = df_notnulls[df_notnulls['fuelType1'] == 'Electricity']
df_elet['displ'].isna().value_counts()

True     118
False      2
Name: displ, dtype: int64

In [99]:
df_elet = df_notnulls[df_notnulls['fuelType1'] == 'Electricity']
df_elet['displ'].value_counts()

0.0    2
Name: displ, dtype: int64

In [100]:
mask_notelet = df_notnulls['fuelType1'] != 'Electricity'
mask_displnan = df_notnulls['displ'].isna()

df_notnulls[mask_notelet & mask_displnan]

barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
21413   13.73375         0.0        0.0        0.0      22      0.0        0   
21414   14.33087         0.0        0.0        0.0      21      0.0        0   

       cityA08U  cityCD  cityE  cityUF  co2  co2A  co2TailpipeAGpm  \
21413       0.0     0.0    0.0     0.0   -1    -1              0.0   
21414       0.0     0.0    0.0     0.0   -1    -1              0.0   

       co2TailpipeGpm  comb08  comb08U  combA08  combA08U  combE  combinedCD  \
21413      370.291667      24      0.0        0       0.0    0.0         0.0   
21414      386.391304      23      0.0        0       0.0    0.0         0.0   

       combinedUF  cylinders  displ          drive  engId  feScore  \
21413         0.0        NaN    NaN  4-Wheel Drive      0       -1   
21414         0.0        NaN    NaN  4-Wheel Drive      0       -1   

       fuelCost08  fuelCostA08 fuelType         fuelType1  ghgScore  \
21413        1400            0  Regular  Regular Gasoline        -1   
21414        1450            0  Regular  Regular Gasoline        -1   

       ghgScoreA  highway08  highway08U  highwayA08  highwayA08U  highwayCD  \
21413         -1         28         0.0           0          0.0        0.0   
21414         -1         27         0.0           0          0.0        0.0   

       highwayE  highwayUF  hlv  hpv     id  lv2  lv4    make     model  \
21413       0.0        0.0    0    0  29472    0    0  Subaru  RX Turbo   
21414       0.0        0.0    0    0  29473    0    0  Subaru  RX Turbo   

      mpgData  phevBlended  pv2  pv4  range  rangeCity  rangeCityA  rangeHwy  \
21413       N        False    0    0      0        0.0         0.0       0.0   
21414       N        False    0    0      0        0.0         0.0       0.0   

       rangeHwyA         trany  UCity  UCityA  UHighway  UHighwayA  \
21413        0.0  Manual 5-spd   27.4     0.0      38.8        0.0   
21414        0.0  Manual 5-spd   26.9     0.0      37.1        0.0   

                VClass  year  youSaveSpend  charge240b  \
21413  Subcompact Cars  1985          -250         0.0   
21414  Subcompact Cars  1985          -500         0.0   

                          createdOn                    modifiedOn  phevCity  \
21413  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013         0   
21414  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013         0   

       phevHwy  phevComb  
21413        0         0  
21414        0         0

In [104]:
mask_subaru = df_notnulls['make'] == 'Subaru'
mask_model = df_notnulls['model'] == 'RX Turbo'
df_notnulls[mask_subaru].groupby('model')['displ'].unique()

model
3 Door 4WD Turbo                                   [1.8]
B9 Tribeca AWD                                     [3.0]
BRZ                                                [2.0]
Baja AWD                                           [2.5]
Brat 4WD                                           [1.8]
Crosstrek AWD                                      [2.0]
Forester AWD                                  [2.5, 2.0]
Hatchback                                     [1.8, 1.6]
Hatchback 4WD                                      [1.8]
Impreza                                            [1.8]
Impreza AWD                         [1.8, 2.2, 2.5, 2.0]
Impreza Sport AWD                                  [2.0]
Impreza Wagon                                      [1.8]
Impreza Wagon AWD                   [1.8, 2.2, 2.5, 2.0]
Impreza Wagon/Outback SPT AWD                 [2.0, 2.5]
Impreza Wagon/Outback Sport                        [2.5]
Impreza Wagon/Outback Sport AWD               [2.5, 2.0]
Justy                    

In [108]:
mask_subaru = df_notnulls['make'] == 'Subaru'
mask_model = df_notnulls['model'] == 'RX Turbo'

df_notnulls['displ_ajust'] = df_notnulls['displ']
df_notnulls.loc[mask_subaru & mask_model, 'displ_ajust'] = 1.8
# df_notnulls.loc[~(mask_subaru & mask_model), 'displ_ajust'] = 2.0

In [112]:
df_notnulls['displ_ajust'].isna().value_counts()

False    37725
True       118
Name: displ_ajust, dtype: int64

In [113]:
df_notnulls['displ_ajust'] = df_notnulls['displ_ajust'].fillna(0)

In [114]:
df_notnulls['displ_ajust'].isna().value_counts()

False    37843
Name: displ_ajust, dtype: int64

## Checking duplicates rows
    
>    - Dropping fully duplicate row
>    - Subset
>    - `.duplicated()`
>    - `.drop_duplicates()`

In [115]:
frutas = pd.DataFrame({
    'fruta':['laranja', 'laranja', 'mamão', 'laranja'],
    'cidade':['Rio de Janeiro', 'Atibaia', 'Campinas', 'Rio de Janeiro']
})

In [116]:
frutas

fruta          cidade
0  laranja  Rio de Janeiro
1  laranja         Atibaia
2    mamão        Campinas
3  laranja  Rio de Janeiro

In [117]:
frutas.iloc[:3]

fruta          cidade
0  laranja  Rio de Janeiro
1  laranja         Atibaia
2    mamão        Campinas

In [123]:
frutas.duplicated()

0    False
1    False
2    False
3     True
dtype: bool

In [122]:
frutas.duplicated(subset=['fruta'])

0    False
1     True
2    False
3     True
dtype: bool

In [127]:
dup2drop = frutas.duplicated(subset=['fruta'], keep='last')
dup2drop

0     True
1     True
2    False
3    False
dtype: bool

In [128]:
frutas[dup2drop]

fruta          cidade
0  laranja  Rio de Janeiro
1  laranja         Atibaia

In [129]:
frutas[dup2drop].index

Int64Index([0, 1], dtype='int64')

In [130]:
frutas.drop(index=frutas[dup2drop].index)

fruta          cidade
2    mamão        Campinas
3  laranja  Rio de Janeiro

In [132]:
frutas.drop_duplicates()

fruta          cidade
0  laranja  Rio de Janeiro
1  laranja         Atibaia
2    mamão        Campinas

In [133]:
frutas.drop_duplicates(subset=['fruta'], keep='last')

fruta          cidade
2    mamão        Campinas
3  laranja  Rio de Janeiro

In [ ]:
#criar method para DataFrames

In [139]:
from pandas.core.base import PandasObject

In [ ]:
def set_description(df, description):
    df.description = description

In [146]:
PandasObject.set_description = set_description

In [147]:
df_notnulls.set_description('oieee')

In [152]:
df_notnulls.description

'oieee'

In [150]:
df.set_description('Este dataFrame tem dados bagunçados de veiculos')

In [151]:
df.description

'Este dataFrame tem dados bagunçados de veiculos'

In [153]:
df2 = df

In [154]:
df2.description

'Este dataFrame tem dados bagunçados de veiculos'